File: myExpenses/src/main/java/org/totschnig/myexpenses/preference/CalendarListPreferenceDialogFragmentCompat.java
package org.totschnig.myexpenses.preference;

import android.content.DialogInterface;
import android.database.Cursor;
import android.database.MatrixCursor;
import android.database.MergeCursor;
import android.os.Build;
import android.os.Bundle;

import com.android.calendar.CalendarContractCompat;

import org.totschnig.myexpenses.MyApplication;
import org.totschnig.myexpenses.R;
import org.totschnig.myexpenses.activity.ProtectedFragmentActivity;
import org.totschnig.myexpenses.provider.DbUtils;

import androidx.appcompat.app.AlertDialog;
import androidx.preference.ListPreference;
import androidx.preference.PreferenceDialogFragmentCompat;

public class CalendarListPreferenceDialogFragmentCompat extends PreferenceDialogFragmentCompat {
  @Override
  protected void onPrepareDialogBuilder(AlertDialog.Builder builder) {
    final ListPreference preference = (ListPreference) getPreference();
    boolean localExists = false;
    Cursor selectionCursor;
    final String value = preference.getValue();
    int selectedIndex = -1;
    String[] projection =
        new String[]{
            CalendarContractCompat.Calendars._ID,
            CalendarContractCompat.Calendars.ACCOUNT_NAME,
            CalendarContractCompat.Calendars.ACCOUNT_TYPE,
            CalendarContractCompat.Calendars.NAME,
            "ifnull(" + CalendarContractCompat.Calendars.ACCOUNT_NAME + ",'') || ' / ' ||" +
                "ifnull(" + CalendarContractCompat.Calendars.CALENDAR_DISPLAY_NAME + ",'') AS full_name"
        };
    Cursor calCursor = null;
    try {
      calCursor = getContext().getContentResolver().
          query(CalendarContractCompat.Calendars.CONTENT_URI,
              projection,
              CalendarContractCompat.Calendars.CALENDAR_ACCESS_LEVEL + " >= " + CalendarContractCompat.Calendars.CAL_ACCESS_CONTRIBUTOR,
              null,
              CalendarContractCompat.Calendars._ID + " ASC");
    } catch (SecurityException e) {
      // android.permission.READ_CALENDAR or android.permission.WRITE_CALENDAR missing
    }
    if (calCursor != null) {
      if (calCursor.moveToFirst()) {
        do {
          if (calCursor.getString(0).equals(value)) {
            selectedIndex = calCursor.getPosition();
          }
          if (DbUtils.getString(calCursor, 1).equals(MyApplication.PLANNER_ACCOUNT_NAME)
              && DbUtils.getString(calCursor, 2).equals(CalendarContractCompat.ACCOUNT_TYPE_LOCAL)
              && DbUtils.getString(calCursor, 3).equals(MyApplication.PLANNER_CALENDAR_NAME))
            localExists = true;
        } while (calCursor.moveToNext());
      }
      if (localExists || Build.VERSION.SDK_INT < Build.VERSION_CODES.JELLY_BEAN) {
        selectionCursor = calCursor;
      } else {
        MatrixCursor extras = new MatrixCursor(new String[]{
            CalendarContractCompat.Calendars._ID,
            CalendarContractCompat.Calendars.ACCOUNT_NAME,
            CalendarContractCompat.Calendars.ACCOUNT_TYPE,
            CalendarContractCompat.Calendars.NAME,
            "full_name"});
        extras.addRow(new String[]{
            "-1", "", "", "",
            getContext().getString(R.string.pref_planning_calendar_create_local)});
        selectionCursor = new MergeCursor(new Cursor[]{calCursor, extras});
      }
      selectionCursor.moveToFirst();
      builder.setSingleChoiceItems(selectionCursor, selectedIndex, "full_name",
          (dialog, which) -> {
            long itemId = ((AlertDialog) dialog).getListView().getItemIdAtPosition(which);
            if (itemId == -1) {
              //TODO: use Async Task Strict Mode violation
              String plannerId = MyApplication.getInstance().createPlanner(false);
              boolean success = !plannerId.equals(MyApplication.INVALID_CALENDAR_ID);
              ((ProtectedFragmentActivity) getActivity()).showSnackbar(
                  success ? R.string.planner_create_calendar_success : R.string.planner_create_calendar_failure);
              if (success) {
                preference.setValue(plannerId);
              }
            } else {
              if (preference.callChangeListener(itemId)) {
                preference.setValue(String.valueOf(itemId));
              }
            }
            CalendarListPreferenceDialogFragmentCompat.this.onClick(dialog,
                DialogInterface.BUTTON_POSITIVE);
            dialog.dismiss();
          });
    } else {
      builder.setMessage("Calendar provider not available");
    }
    builder.setPositiveButton(null, null);
  }

  @Override
  public void onDialogClosed(boolean b) {
    //nothing to do since directly handled in onClickListener of SingleChoiceItems
  }

  public static CalendarListPreferenceDialogFragmentCompat newInstance(String key) {
    CalendarListPreferenceDialogFragmentCompat fragment = new CalendarListPreferenceDialogFragmentCompat();
    Bundle bundle = new Bundle(1);
    bundle.putString(ARG_KEY, key);
    fragment.setArguments(bundle);
    return fragment;
  }
}

File: myExpenses/src/main/java/org/totschnig/myexpenses/preference/CalendarListPreferenceDialogFragmentCompat.kt
package org.totschnig.myexpenses.preference

import android.content.DialogInterface
import android.database.Cursor
import android.database.MatrixCursor
import android.database.MergeCursor
import android.os.Bundle
import android.provider.CalendarContract
import androidx.appcompat.app.AlertDialog
import androidx.preference.ListPreference
import androidx.preference.PreferenceDialogFragmentCompat
import org.totschnig.myexpenses.R
import org.totschnig.myexpenses.activity.ProtectedFragmentActivity
import org.totschnig.myexpenses.injector
import org.totschnig.myexpenses.provider.INVALID_CALENDAR_ID
import org.totschnig.myexpenses.provider.PLANNER_ACCOUNT_NAME
import org.totschnig.myexpenses.provider.PLANNER_CALENDAR_NAME
import org.totschnig.myexpenses.provider.PlannerUtils
import org.totschnig.myexpenses.provider.requireString
import javax.inject.Inject

class CalendarListPreferenceDialogFragmentCompat : PreferenceDialogFragmentCompat() {

    @Inject
    lateinit var plannerUtils: PlannerUtils

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        injector.inject(this)
    }

    override fun onPrepareDialogBuilder(builder: AlertDialog.Builder) {
        val preference = preference as ListPreference
        var localExists = false
        val selectionCursor: Cursor
        val value = preference.value
        var selectedIndex = -1
        val projection = arrayOf(
            CalendarContract.Calendars._ID,
            CalendarContract.Calendars.ACCOUNT_NAME,
            CalendarContract.Calendars.ACCOUNT_TYPE,
            CalendarContract.Calendars.NAME,
            "ifnull(" + CalendarContract.Calendars.ACCOUNT_NAME + ",'') || ' / ' ||" +
                    "ifnull(" + CalendarContract.Calendars.CALENDAR_DISPLAY_NAME + ",'') AS full_name"
        )
        val calCursor = try {
            requireContext().contentResolver.query(
                CalendarContract.Calendars.CONTENT_URI,
                projection,
                CalendarContract.Calendars.CALENDAR_ACCESS_LEVEL + " >= " + CalendarContract.Calendars.CAL_ACCESS_CONTRIBUTOR,
                null,
                CalendarContract.Calendars._ID + " ASC"
            )
        } catch (e: SecurityException) {
            null
            // android.permission.READ_CALENDAR or android.permission.WRITE_CALENDAR missing
        }
        if (calCursor != null) {
            if (calCursor.moveToFirst()) {
                do {
                    if (calCursor.getString(0) == value) {
                        selectedIndex = calCursor.position
                    }
                    if (calCursor.requireString(1) == PLANNER_ACCOUNT_NAME && calCursor.requireString(
                            2
                        ) == CalendarContract.ACCOUNT_TYPE_LOCAL && calCursor.requireString(3) == PLANNER_CALENDAR_NAME
                    ) localExists = true
                } while (calCursor.moveToNext())
            }
            selectionCursor = if (localExists) {
                calCursor
            } else {
                val extras = MatrixCursor(
                    arrayOf(
                        CalendarContract.Calendars._ID,
                        CalendarContract.Calendars.ACCOUNT_NAME,
                        CalendarContract.Calendars.ACCOUNT_TYPE,
                        CalendarContract.Calendars.NAME,
                        "full_name"
                    )
                )
                extras.addRow(
                    arrayOf(
                        "-1", "", "", "",
                        requireContext().getString(R.string.pref_planning_calendar_create_local)
                    )
                )
                MergeCursor(arrayOf(calCursor, extras))
            }
            selectionCursor.moveToFirst()
            builder.setSingleChoiceItems(
                selectionCursor, selectedIndex, "full_name"
            ) { dialog: DialogInterface, which: Int ->
                val itemId = (dialog as AlertDialog).listView.getItemIdAtPosition(which)
                if (itemId == -1L) {
                    //TODO: use Async Task Strict Mode violation
                    //noinspection MissingPermission
                    val plannerId: String = plannerUtils.createPlanner(false)
                    val success = plannerId != INVALID_CALENDAR_ID
                    (activity as ProtectedFragmentActivity?)!!.showSnackBar(
                        if (success) R.string.planner_create_calendar_success else R.string.planner_create_calendar_failure
                    )
                    if (success) {
                        preference.value = plannerId
                    }
                } else {
                    if (preference.callChangeListener(itemId)) {
                        preference.value = itemId.toString()
                    }
                }
                onClick(
                    dialog,
                    DialogInterface.BUTTON_POSITIVE
                )
                dialog.dismiss()
            }
        } else {
            builder.setMessage("Calendar provider not available")
        }
        builder.setPositiveButton(null, null)
    }

    override fun onDialogClosed(b: Boolean) {
        //nothing to do since directly handled in onClickListener of SingleChoiceItems
    }

    companion object {
        fun newInstance(key: String?): CalendarListPreferenceDialogFragmentCompat {
            val fragment = CalendarListPreferenceDialogFragmentCompat()
            val bundle = Bundle(1)
            bundle.putString(ARG_KEY, key)
            fragment.arguments = bundle
            return fragment
        }
    }
}
